include/unifex/bulk_schedule.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/get_execution_policy.hpp>
+#include <unifex/execution_policy.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _bulk_schedule {
+
+template<typename Integral, typename Receiver>
+struct _schedule_receiver {
+    class type;
+};
+
+template<typename Integral, typename Receiver>
+using schedule_receiver = typename _schedule_receiver<Integral, Receiver>::type;
+
+template<typename Integral, typename Receiver>
+class _schedule_receiver<Integral, Receiver>::type {
+public:
+    template<typename Receiver2>
+    explicit type(Integral count, Receiver2&& r)
+    : count_(std::move(count))
+    , receiver_((Receiver2&&)r)
+    {}
+
+    void set_value()
+        noexcept(is_nothrow_value_receiver_v<Receiver> &&
+                 is_nothrow_next_receiver_v<Receiver, Integral>) {
+        using policy_t = decltype(get_execution_policy(receiver_));
+        if constexpr (is_one_of_v<policy_t, unsequenced_policy, parallel_unsequenced_policy>) {
```

Comment:
> (If we didn't have that 'property system' thing I wouldn't be concerned)

Can you elaborate?

Reply:
If an algorithm must declare that it needs stronger concurrency guarantees to do its job than the underlying executor, it has to do that by creating its own executor and returning different values from properties (if I understand correctly). But we don't know all properties that must be edited to say "no I really need sequenced here." In current P0443 that happens between blocking_t and bulk_guarantee_t -- if bulk_execute hasn't been customized and blocking_t is not 'always', then the only possible bulk_guarantee_t the fallback can provide is 'unsequenced'. With the parallel algorithm execution policies I could see a similar problem/conflict between properties being possible, leading to a case where the algorithm says it wants ordered/sequenced behavior but doesn't override every property other parts of the system use when making such determinations.

That is to say, the current algorithm execution policies don't say anything about the executor. They describe guarantees the user is making about their element access functions at the call site. In that sense involving that in any way as a property of the executor might be backwards.... just because the user's element access functions tolerate parallel_unsequenced_t doesn't mean the algorithm can tolerate an executor that provides such weak guarantees. So the algorithm needs to be able to adapt the executor into one providing stronger guarantees (e.g. with require()). That means we can't ever add a property that some other part of the system might interpret in a way that breaks other properties' guarantees, so if we had already shipped P0443 we would not be able to add this execution policy property because it conflicts with `bulk_guarantee_t`.

Reply:
The `get_execution_policy()` query is there to allow a many-sender the ability to query the constraints on how it may call the the `set_next()` method of the receiver. The builtin policies provide constraints along two orthogonal axes: 'allow concurrent calls' (par, par_unseq) and 'allow interleaved calls' (unseq, par_unseq).

With these policies a receiver is not able to dictate that a sender _must_ call it concurrently or interleaved. It is always valid for a given sender to call it sequentially from a single thread.

One gap that is currently missing is the ability for an algorithm to query whether a given sender _may_ call the receiver concurrently or interleaved. An algorithm may choose to provide a receiver that takes a more efficient strategy if it knows it will not be called concurrently (eg. by avoiding use of atomics).

The algorithm/many-receiver and executor/many-sender can negotiate between them an execution strategy by allowing the algorithm to query the *capabilities* of a many-sender, and thus choose the most appropriate receiver to pass to the `connect()` call, and to allow a many-sender to query the *constraints* that the many-receiver places on calls to `set_next()`, so that it can choose an appropriate execution strategy.

There will likely also be other queries that are possible. eg. the 'adjacency' property proposed in [P2155R0](https://wg21.link) could be implemented as a `get_adjacency_hint()` CPO customisable on a many-receiver that a many-sender can query to refine its strategy for calling `set_next()`.

Reply:
>With these policies a receiver is not able to dictate that a sender must call it concurrently or interleaved

I'm saying that algorithms need the opposite: they need to be able to dictate that they must **not** be called interleaved in places, even if the user passed an executor with the unseq or par_unseq property when the algorithm was called.

For example our implementation does not have a parallel `merge()` so we need to say "too bad, you get serial".

Reply:
> I'm saying that algorithms need the opposite: they need to be able to dictate that they must not be called interleaved in places, even if the user passed an executor with the unseq or par_unseq property when the algorithm was called.

Yes. This is what we have currently implemented and what `get_execution_policy(receiver)` gives you.

It lets the receiver return, say, `std::par`, which indicates it is ok with being called concurrently, or sequentially, but is not allowed to be called interleaved. An executor that has the ability to call a receiver interleaved will need to fall back to a non-interleaved execution strategy. ie. the receiver has put a constraint on the valid execution-strategies that the sender can use.

Reply:
Right, that's certainly something we can do, but there are already properties in P0443 (like blocking_t) that could contradict that setting. That's what I meant by:

>(If we didn't have that 'property system' thing I wouldn't be concerned)

e.g. if you say execution policy seq but blocking_t says nonblocking or bulk_guarantee_t says unsequenced the result is semantically broken. So we need to override all of these properties that might indicate some degree of parallelism if the algorithm can't tolerate that and I'm not sure that's reasonable.

Reply:
I'm not sure I agree with the current design of the 'blocking_t' property as it stands in P0443.

In libunifex we have a `blocking()` query on a sender that allows you to query the disposition of that sender's completion: namely whether it will complete inline before start() returns, completes on some other thread before start() returns, or will never complete inline on the current thread. These can be used to apply certain optimisations in algorithms (like `sync_wait()`, or `submit()`) to avoid heap-allocation or synchronisation.

But I'm not sure it makes sense to be able to `require()` that an executor be `blocking::always`. While it may have been a necessity in a world where you only had an eager `execute()` function. But once you have lazy sender-based operations and schedulers with a `schedule()` basis operation, you can just call `sync_wait()` to block waiting for any async operation to complete.

For the `bulk_guarantee_t` property: a given many-sender, eg. that is backed by a GPU, may _only_ support unsequenced execution. In this case it would be perfectly fine for that sender's `connect()` to be constrained such that it requires receivers passed to have a `get_execution_policy()` that is compatible with the requirements. If you try to give it a receiver that does not allow unsequenced execution then this would manifest as a compile-error.

This would limit the composability of these many-senders with some kinds of algorithms, though, so I'm not yet sold on whether this is "A Good Thing" to be adding.

Reply:
>For the bulk_guarantee_t property: a given many-sender, eg. that is backed by a GPU, may only support unsequenced execution.

Then there will be algorithms that can't be used with that executor. The current parallel algorithms design is 'add a parallel overload unless it is blindingly obvious that no parallel version is possible', on the grounds that implementations where a parallel version is impossible can fall back to providing a serial version. There are still a lot of algorithms for which we don't know a parallel version faster than the serial version, including some (like `generate` and `partial_sort`) for which we think no parallel version is possible and the parallel overloads were added in error. See https://github.com/microsoft/STL/blob/0cdf5fbfac83650bf6b404a754be415493466643/stl/inc/yvals_core.h#L285-L311